src/core/dbus-job.c

```diff
@@ -286,14 +287,45 @@ static int send_removed_signal(sd_bus *bus, void *userdata) {
                         &m,
                         "/org/freedesktop/systemd1",
                         "org.freedesktop.systemd1.Manager",
-                        "JobRemoved");
+                        "JobRemoved2");
         if (r < 0)
                 return r;
 
         r = sd_bus_message_append(m, "uoss", j->id, p, j->unit->id, job_result_to_string(j->result));
         if (r < 0)
                 return r;
 
+        r = sd_bus_send(bus, m, NULL);
+        if (r < 0)
+                return r;
+
+        /* Only send out legacy signal if there is at least one subscriber who asked for it. */
+        if (j->manager->n_legacy_subscribers == 0)
+                return 0;
+
+        m = sd_bus_message_unref(m);
+
+        r = sd_bus_message_new_signal(
+                        bus,
+                        &m,
+                        "/org/freedesktop/systemd1",
+                        "org.freedesktop.systemd1.Manager",
+                        "JobRemoved");
+        if (r < 0)
+                return r;
+
+        /* The old JobRemoved signal never returns 'skipped' as a result, unless the job was about a device
+         * unit, for backward-compatibility reasons. */
+        JobResult result;
+        if (j->result == JOB_SKIPPED && j->unit->type != UNIT_DEVICE)
```

Comment:
hmm, where was that conditionalization on devices? i missed that?

Reply:
It always returned 'skipped' when the  unit doesn't have a `start()` method, which means device unit:

https://github.com/systemd/systemd/blob/ff7e7c2b3a2b4367e0a03a51b7d714dfe0837abb/src/core/unit.c#L1916
https://github.com/systemd/systemd/blob/ff7e7c2b3a2b4367e0a03a51b7d714dfe0837abb/src/core/unit.c#L1842
https://github.com/systemd/systemd/blob/ff7e7c2b3a2b4367e0a03a51b7d714dfe0837abb/src/core/job.c#L894

Reply:
uh, oh, so "skipped" actually means something entirely different than "condition failed". It means "you asked me for an operation i can not execute for you on this unit type, but all is good, so i skipped it".

we need a new JobResult for indicating skipped conditions. "conditioned" or so?

Reply:
Ah, fuck, ExecCondition= has already squatted on this result, too. what a mess... see job.h, the JobResult definition.

we need to clean that up.


i.e. so what about this. We introduce:

JOB_DONE_CONDITION + JOB_DONE_EXEC_CONDITION or so as new "special flavours" of JOB_DONE. That way clients can look for "success conditions" via `streq(result, "done") || startswith(result, "done-")`, in a generic fashion.

Then in the compat JobRemoved signal mangle JOB_DONE_EXEC_CONDITION into JOB_SKIPPED and
JOB_DONE_CONDITION into JOB_DONE. And in the new JobRemovedEx we'd instead pass them down in their proper names.

With that in place things  should be reasonably clean again and one result will mean exactly one thing again.

Reply:
Aside from that "on this unit type", which is an implementation detail, to me it seems a description that fits pretty well for this case though - the operation could not be executed, but the reason it could not be executed is within the parameters set by the system configuration, so it's all good, no?

It seems to fit with the original description of the job result even when it was introduced: https://github.com/systemd/systemd/commit/d68201e9aa5e9ebd6085b1bb8892c42e9d20be75

Reply:
> JOB_DONE_CONDITION + JOB_DONE_EXEC_CONDITION or so as new "special flavours" of JOB_DONE. That way clients can look for "success conditions" via `streq(result, "done") || startswith(result, "done-")`, in a generic fashion.

Why are different return values for Condition/ExecCondition needed? The latter is just a free-form of the former, no? In which situation should a caller do something different if a condition caused a skip vs if an execcondition caused a skip?

Reply:
I agree. It sounds nicer to have the same result for both.

Reply:
> Why are different return values for Condition/ExecCondition needed? The latter is just a free-form of the former, no? In which situation should a caller do something different if a condition caused a skip vs if an execcondition caused a skip?

I think they are quite different, because the exec one actually executed code, had a blocking runtime and so on. 

But the main reason is because you want to mangle them differently later on for the legacy JobRemoved. You can't do that if you don't distinguish them. i.e. one should be mangled into JOB_SKIPPED and the other into JOB_DONE, see above. Hence, if we have to internally distinguish them I think it makes a lot of sense to also dsitinguish them in the modern external interface too.